### ISSUE TITLE
Bot builders interfere when real players are building

### ISSUE THREAD

**Issue by slipher** [Unvanquished/Unvanquished#2852]
When the bot building feature was introduced, it was promised that they will not try to build if any real players are currently building. However, if I add some bots and go granger on Parpax and deconstruct all but one egg, after a short time a bot will spawn granger even though I'm still a granger.

**Comment by sweet235** [comment]
Not only on parpax, and not only alien bots. This is specified in the default behavior, here: https://github.com/Unvanquished/Unvanquished/blob/master/pkg/unvanquished_src.dpkdir/bots/default.bt#L11 (and below for humans). The condition is:

```numUsersInTeam == 0 || numOurBuildings( E_A_SPAWN ) < 2```

The idea is simply that if a team lost all their spawns (but one), users on their team will frequently be incapacitated in the spawn queue. So one bot tries to obtain the ability to build, and builds until the team has two spawns. Two spawns are universally accepted to be the sensible minimum.

The situation described here is rather artificial. If a user deconstructs all spawns (but one), their building rights will quickly be revoked by their team. Even when moving the base, it is customary to maintain two spawns at all times. The more frequent reason to trigger the bots doing this is an unexpectedly successful attack resulting in the loss of most spawns.

In fact, the bots might build more than one or two spawns in this case. They might also build the reactor/overmind, and a drill/leech in order to try to be able to build the spawns.

> it was promised that they will not try to build if any real players are currently building

They will not build if any user is on their team. This is the only exception. I added it after quite a few people had suggested this.

How to proceed? I see these options:

- Life with it, and see how it turns out in real games
- Require the additional condition that all users are indeed in the spawn queue
- Remove it (removing lines 11-12 and 33-34 removes this)

**Comment by sweet235** [comment]
Maybe simply removing this exception is the most intuitive thing to do from a user perspective. We would still be able to make the bots build if there are no spawns left, by giving the command `/tactic build` (provided this is in `g_tacticBehaviors`).

**Comment by sweet235** [comment]
There is a fourth option:

- Require the additional condition that no user on their team is currently a builder

But then, a few days ago, I was watching a game with a malicious deconner. When switched teams and started to remove spawns again, I was away from my computer, and people started to cry for an admin intervention. When I came back two minutes later, the user was still removing spawns and a bot was still rebuilding them. Otherwise, the game would have been over.

**Comment by DolceTriade** [comment]
I think it's reasonable to have bots force build at least two spawns regardless of whether there is a builder or not. We can always revert malicious builders anyways... inept builders and trolls on the other hand...let's hope the bots save us :D

**Comment by sweet235** [comment]
Yes. Our bots are a little more useful than they used to be. But for them to be effective, some commands by users are required. I thought that losing spawns is so severe a problem that they could should handle this without any user intervention.

There is a fifth option:

- Only make them build one spawn

**Comment by slipher** [comment]
Now I remember reviewing that PR. So yeah I'm writing a lot of stupid comments. 

> But then, a few days ago, I was watching a game with a malicious deconner. When they switched teams and started to remove spawns again, I was away from my computer, and people started to cry for an admin intervention.

Don't they know how to use vote kick?

Anyway building when there is still a remaining spawn AND another builder seems overly aggressive. Bots can react instantly to changes in conditions so the flesh-and-blood builder might not even have a chance to place a spawn himself.

**Comment by sweet235** [comment]
> Don't they know how to use vote kick?

Some know it, some do not. Also: votes are rather slow. The actual delay consists of:

- Time until someone notices the malicious user
- Time until they figure out how to start the vote
- Time until enough people have voted, or 30 seconds timeout

Some know about the denybuild vote, but might need 10 or 20 second to issue it (there is no GUI). In practice, the whole delay can be over 60 seconds.
